src/amdgpu_pptable/tui.py

```diff
@@ -0,0 +1,281 @@
+            self.edit()
+        else:
+            app = self.parentApp
+            app.file_path = file_path
+
+            if self.mode == MODE_OPEN:
+                app.load(app.file_path)
+            else:
+                app.save(app.file_path)
+
+            self.editing = False
+
+    def on_cancel(self):
+        self.editing = False
+
+
+class AMDPPTTreeForm(npyscreen.FormBaseNewWithMenus):
+    MENU_KEY = "^X"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        widget = self._NMDisplay._DisplayArea._menuListWidget
+        widget.add_handlers(
+            {ord("x"): widget.h_exit_escape, curses.ascii.ESC: widget.h_exit_escape}
+        )
+
+    def create(self):
+        self.main_menu = self.add_menu(
+            name="Main Menu", shortcut=self.__class__.MENU_KEY
+        )
+        self.main_menu.addItemsFromList(
+            [
+                ("Open...", self.parentApp.show_load_form, "^O"),
+                ("Revert", self.parentApp.revert, "^R"),
+                ("Save...", self.parentApp.show_save_form, "^S"),
+                ("Exit Application", self.parentApp.exit_application, "^K"),
+            ]
+        )
+        self.tree_widget = self.add(
+            AMDPPTTreeWidget, color="NO_EDIT", highlight_color="STANDOUT"
+        )
+        self.tree_widget.annotation_padding = 32
+
+
+# extend TreeData class to include row builder
+class AMDPPTTreeData(npyscreen.TreeData):
+    @property
+    def record_type(self):
+        return self._record_type if hasattr(self, "_record_type") else None
+
+    @record_type.setter
+    def record_type(self, record_type):
+        self._record_type = record_type
+
+    @staticmethod
+    def type_label(type):
+        if hasattr(type, "wrapped_type"):
+            return AMDPPTTreeData.type_label(type.wrapped_type)
+        elif issubclass(type, ctypes.Array):
+            return f"{type._type_.__name__}[{type._length_}]"
+        else:
+            return type.__name__
+
+    @staticmethod
+    def short_type_label(type):
+        return AMDPPTTreeData.type_label(type).replace("struct__", "")
+
+    def append_row(self, name, obj, edit_func=None):
+        tree_data = self.new_child(expanded=False, annotation=name)
+        obj_type = type(obj)
+
+        if isinstance(obj, ctypes.Structure):
+            tree_data.set_content(self.short_type_label(obj_type))
+
+            for field_name, field_value in obj_type._fields_:
+                tree_data.append_row(
+                    field_name,
+                    getattr(obj, field_name),
+                    functools.partial(setattr, obj, field_name),
+                )
+
+        elif isinstance(obj, ctypes.Array):
+            tree_data.set_content(self.short_type_label(obj_type))
+
+            for i in range(len(obj)):
+                tree_data.append_row(
+                    f"{name}[{i}]", obj[i], functools.partial(operator.setitem, obj, i)
+                )
+
+        else:
```

Comment:
BTW, there are also a few MutableMappings (ArrayDictAdapters https://github.com/amezin/amdgpu-pptable/blob/master/src/amdgpu_pptable/smu_v11_0_pptable_navi10.py#L25, for Navi only):
https://github.com/amezin/amdgpu-pptable-editor-qt/blob/master/src/amdgpu_pptable_qt/gui.py#L34